### ISSUE TITLE
find() returning completely wrong data/schema

### ISSUE THREAD

**Issue by ffxsam** [jagi/meteor-astronomy#495]
I don't know if anything will come of this, since I can't seem to reproduce this. But I thought I'd report it anyway, as it's such a catastrophic problem that I've had to remove Astronomy from my project.

The results of a `.find().fetch()` of one item is mixing properties with a completely different data model. I'm passing the results of `Insights.find().fetch()` into a React component.. yet it has NONE of the properties an Insight has, and instead has the same properties a Tag model does. Yet in the Meteor container, if I just do a `console.log(Insights.find().fetch())`, then the right data is passed into the React component, which doesn't make sense.

So for example, this is the schema for `Insight`:

``` js
export const Insight = Class.create({
  name: 'Insight',
  collection: Insights,
  behaviors: ['timestamp', 'softremove'],
  secured: false,
  fields: {
    attachments: {
      type: [String], // array of attachment IDs
      default() {
        return []
      },
    },
    avgRating: {
      type: Number,
      default() {
        return 0;
      },
    },
    comments: {
      type: [Comment],
      default() {
        return [];
      },
    },
    coverAttachment: {
      type: String,
      optional: true,
      default: '',
    },
    description: { type: String, optional: true },
    lastCommentedAt: {
      optional: true,
      type: Date,
    },
    orgId: String,
    ownerId: {
      type: String,
      default() {
        return Meteor.userId();
      },
    },
    ratings: {
      type: [Object],
      default() {
        return []
      },
    },
    recommendations: { type: String, optional: true },
    related: {
      type: [Object],
      default() {
        return []
      },
    },
    tagIds: {
      type: [String],
      default() {
        return []
      },
    },
    title: String,
  },

  methods: {
    ...attachmentMethods,
    ...commentMethods,
    ...ratingMethods,
    ...tagMethods,
  },
});
```

But when I `console.log` the result of a `find` in Chrome, I see this:

![image](https://cloud.githubusercontent.com/assets/12532733/18177628/e857db02-702f-11e6-9f3b-39c15ed1db0b.png)

`label` is not even a field of `Insight`—in fact it's a field in `Tag`. And if I do `instanceof Tag` on this very same object, it returns `true`, when clearly the results are from `Insights.find().fetch()`.

In my Meteor container's return statement, if I change `availableTags: Tag.find({}, alphaSort).fetch()`, to use `Tags` (Mongo, directly) instead, everything works great. So just by using Astronomy's find method, it's screwing up another model's schema somehow.

Any ideas what might be causing this?

**Comment by lukejagodzinski** [comment]
I assume that there is no inheritance between `Tag` and `Insight`. It seems to be a problem with references.

It's hard to tell without the reproduction where is a problem. If there is a difference in what `Insight.find().fetch()` and `Insights.find().fetch()` return just by calling them in the browser's console then it seems to be a bug in Astronomy but if it returns difference values but inside some other tool then it might mean that a problem is in your code. Some libraries don't like objects created using classes and they require plain objects to work. So you may need using `doc.raw()` method or turn off transformation `Insights.find({}, {transform: null}).fetch()`.

**Comment by ffxsam** [comment]
No, that's the weird thing. In the console, `Insights.find().fetch()` and `Insight.find().fetch()` return what I'd expect. It's the Meteor container where it's not passing the right things. E.g.:

``` js
const MeteorContainer = createContainer(({ actingOrgId, insightId }) => {
  const addAttachment = (args, callback) => {
    _addAttachment.call(args, callback);
  };
  const alphaSort = { sort: { title: 1 } };
  const handles = [
    Meteor.subscribe('allUsers'),
    Meteor.subscribe('insights', { orgId: actingOrgId }),
    Meteor.subscribe('tests', { orgId: actingOrgId }),
    Meteor.subscribe('notes', { orgId: actingOrgId }),
  ];

  return {
    addAttachment,
    attachments: Attachment.find().fetch(),
    availableTags: Tags.find({}, alphaSort).fetch(),
    createTag,
    dataReady: subscriptionsReady(handles),
    insight: Insight.findOne(insightId),
    insights: Insight.find({}, alphaSort).fetch(),
    otherItems: [
      ...Test.find({}, alphaSort).fetch(),
      ...Note.find({}, alphaSort).fetch(),
    ],
    userId: Meteor.userId(),
    users: Meteor.users.find().fetch(),
  }
}, InsightDetail);
```

So the `otherItems` prop contains an array of items, which _some_ of the correct data in it (as long as the properties overlap with `Insight`). But `Test` has a field `hypothesis` which isn't showing up, the `.constructor.className` of all the items in `otherItems` returns `"Insight"`.

Do you have any ideas why this might be happening?

**Comment by ffxsam** [comment]
> I assume that there is no inheritance between Tag and Insight.

Correct, none of these are connected in any way.

**Comment by ffxsam** [comment]
BTW, changing the above to this, fixes it:

``` js
    otherItems: [
      ...Test.find({}, { ...alphaSort, transform: null }).fetch(),
      ...Note.find({}, { ...alphaSort, transform: null }).fetch(),
    ],
```

Of course, now I'm unable to use Astro methods on these objects. So there's something wrong with the transformation process. I'm happy to clone this and use a local copy of Astronomy, and mess with the source code to see if I can figure out what's going on. Could you maybe save me some time (since I'm not familiar with the code) and please point out where the data transformation occurs? Then I'll take it from there.

**Comment by ffxsam** [comment]
Never mind, found the relevant code! Ok, I've narrowed it down to a problem within `transformToClass`. At the beginning, `rawDoc` is correct, and on [this line](https://github.com/jagi/meteor-astronomy/blob/v2/lib/modules/storage/utils/transform_to_class.js#L32) it's returning the totally wrong document.

**Comment by ffxsam** [comment]
@jagi Ok, this is super weird. I changed one of my lines above to this:

``` js
 ...Test.find({ hypothesis: /[A-Z]/ }, alphaSort).fetch()
```

just so I could easily track it in your code. Here's the weird part I can't figure out. In your `find.js` I modified it:

``` js
function createMethod(methodName) {
  return function(selector, options) {
    let Class = this;
    let Collection = Class.getCollection();

    console.error('find() called:', selector, options);
```

And it's printing this out:

![image](https://cloud.githubusercontent.com/assets/12532733/18188251/9608d12c-7065-11e6-812c-bac2eef8eaab.png)

Where is this transform func coming form? I'm definitely not passing it in my `find()` call! And this function is responsible for using the wrong class (Insight) to transform the data.

**Comment by ffxsam** [comment]
I still can't pinpoint it, but it seems like `option.transform` is getting set to a function to transform a raw doc to an `Insight` class, and then the next time `find` runs (even if it's on a `Test`), it uses the same `option.transform` as before, which was for an `Insight`.

**Comment by lukejagodzinski** [comment]
Ok to just go through it faster because I don't have time to analyse all the code examples you wrote. After skimming through the PR I'm assuming that what you were doing is something like:

``` js
var options = {
  /* some options here */
};
Insight.find({}, options);
Tag.find({}, options);
```

So because of not cloning the `options` object by the `find` method you were reusing `transform` option from the previous call? The `transform` option is set by Astronomy if you don't provide your own. That way it can construct Astronomy class instances. And I agree that it's a bug because it's not cloning object.

However, I don't like the name `mutableOptions`. You could just change 2 lines instead of chaning name of var in every possible place :). Can you update that? And I will merge PR :). Thanks for finding this bug!

``` js
options = _.defaults({}, options, {
  defaults: true,
  children: true
});
```

**Comment by ffxsam** [comment]
It's extremely bad practice to make assignments to a function argument, which is why I have a different name `mutableOptions`.. which is a bit wordy, but at least it explains to any dev looking at the code that these options are mutable, while the function's `options` argument should remain untouched.

I'm happy to choose another name, but I won't assign a value to `options` because that's what got us here in the first place. :)

**Comment by ffxsam** [comment]
Whoa... ok, I guess my JavaScript skills are at maybe 80% instead of 95% ;) I tried your code, and sure enough it works fine. I guess I don't quite understand how this:

``` js
options = _.defaults({}, options, {
  defaults: true,
  children: true
});
```

is different from the old buggy code:

``` js
    options = options || {};
    _.defaults(options, {
      defaults: true,
      children: true
    });
```

Both are writing directly to `options`. What am I missing?
I'll go ahead and revise my commit, rebase, and push.

**Comment by ffxsam** [comment]
Ah, I got it. Because of this line:

``` js
options = options || {};
```

So since `options` is being passed as an argument, we're essentially doing `options = options` which doesn't create a new object, and then proceeds to use `_.defaults` to set properties into that object. Whereas with `options = _.defaults({}, options, { .. })` we're assigning to a function argument, though it's a brand new object so the reference is broken, and it does no harm to the variable passed in.

I updated the PR, take a look!
